lib/rails/generators/mongoid/config/config_generator.rb

```diff
@@ -25,9 +25,50 @@ def app_name
       end
 
       def create_config_file
-        template 'mongoid.yml', File.join('config', "mongoid.yml")
+        template 'mongoid.yml', File.join('config', 'mongoid.yml')
       end
 
+      def create_initializer_file
+        template 'mongoid.rb', File.join('config', 'initializers', 'mongoid.rb')
+      end
+
+      private
+
+      # Extracts the available configuration options from the Mongoid::Config
+      # source file, and returns them as an array of hashes.
+      #
+      # @param [ Integer ] indent How many spaces each comment ought to
+      #   be re-indented.
+      # @param [ true | false ] include_deprecated Whether deprecated options
+      #   should be included in the list.
+      #
+      # @return [ Array<Hash<desc: [String], name: [String], default: [String],
+      #   deprecated: [nil | Integer]>> ] the array of hashes representing
+      #   each defined option, in alphabetical order by name.
+      def mongoid_config_options(indent: 2, include_deprecated: false)
```

Comment:
I like everything else this PR, but the idea of parsing source code with a regex to generate another source code scares me. It seems very fragile.
Is there any other reasonable way to do so?

Reply:
Strongly agreed. Perhaps the code generation could be extracted to a Rake task.

Reply:
In defense of the current approach, it doesn't generate _code_, it generates _comments_. Thus, in the worst case, it generates nonsense that doesn't actually break anything. It is unlikely to do so, though, unless we drastically change the format of `mongoid/config.rb`, in which case, the corresponding test (in app_spec.rb) would likely also break and alert us to that fact.

However, here other options we could consider:

- We could **nix the automation altogether** and just duplicate the documentation. This is what is being done in the released versions currently, but it is fragile in another signficant way: it requires the documentation to be maintained in two places (mongoid/config.rb AND the generator template file), and it is very easy to forget about the template file when adding or updating configuration options. I do not recommend this approach.
- We could **use a rake task** to parse `mongoid/config.rb` ahead of time, scraping the documentation and building the template. This does nothing to address the fragility concern, though, because it's doing the exact same work, merely adding a new build step (which is itself prone to be forgotten).
- We could **store the documentation in a separate location**, as @johnnyshields suggested, and then use a rake task to generate `mongoid/config.rb` by combining the doc file with a separate code template. I don't like this approach either, though, because it adds a new build step that must be remembered every time the code template changes, or the docs change. Further, what happens if a developer forgets that `mongoid/config.rb` is autogenerated and edits it directly? A huge warning comment at the top of the file might help (e.g. "THIS IS AUTOGENERATED"), with instructions on where to go to find the template and the docs file, but I think that just adds a lot of complexity on top of something that really doesn't need to be complicated.
- We could **make the comments themselves executable**, like the `desc` directive in Rake tasks. Then, `Mongoid::Config` could be queried directly for all options and their descriptions. This sounds wonderful...except that it breaks our existing doc generator pipeline (which expects rdoc-style comments).
- We could **extract the `mongoid_config_options` method** to a new module, `Mongoid::Config::Introspection`. This would allow us to write more rigorous tests around the scraper, and would hopefully build more confidence in it.

I prefer the last option (`Mongoid::Config::Introspection`) since it better isolates the functionality, gives us an opportunity to write focused tests for it, and makes it clearer that there is a relationship between the format of the comments in `Mongoid::Config`, and the code that scrapes them.

Reply:
@jamis how about we stick with the existing manual approach for this PR, then split this refactor out into a new ticket/PR? I think you're absolutely right to consolidate the options (esp. so we can deprecate them easier), it just feels beyond the narrow scope of this particular change.

Reply:
Good point, @johnnyshields -- the PR has evolved and I think you're right, the consolidation of the options in the yaml file is out of scope now. I'll split that out and treat it separately.